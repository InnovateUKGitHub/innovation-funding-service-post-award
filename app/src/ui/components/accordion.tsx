import * as React from "react";

interface Props {
  title: string;
  content: JSX.Element;
}

interface State {
  accordionOpen: boolean;
}

export class Accordion extends React.Component<Props, State> {

  constructor(props: Props) {
    super(props);
    this.state = {
      accordionOpen: false
    };
  }

  render() {
    const styles = {
      buttonStyle: {
        background: "none",
        border: "none"
      }
    };

    return (
      <div className="govuk-accordion__section">
        <div className="govuk-accordion__section-header" onClick={() => this.setState({accordionOpen: !this.state.accordionOpen})}>
          <button style={styles.buttonStyle} className="govuk-heading-m govuk-!-margin-bottom-3 govuk-!-margin-top-3">{this.props.title}</button>
          <img className="govuk-!-padding-right-2" src={this.state.accordionOpen ? "/assets/images/icon-minus.png" : "/assets/images/icon-plus.png"}/>
        </div>
        {this.state.accordionOpen && <div className="govuk-accordion__section-panel">
          {this.props.content}
        </div>}
      </div>
    );
  }
}
