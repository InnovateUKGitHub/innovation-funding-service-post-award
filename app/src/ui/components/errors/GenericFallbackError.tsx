import { ErrorPayload } from "@shared/create-error-payload";
import { PageTitle } from "@ui/features/page-title";
import { useStores } from "@ui/redux";

import { Page } from "../layout/page";
import { Section } from "../layout/section";
import { H2 } from "../typography";
import { ExternalLink } from "../renderers";
import { SimpleString } from "../renderers/simpleString";
import { Content } from "../content";

export type GenericFallbackErrorProps = ErrorPayload["params"];

export const GenericFallbackError = ({ errorStack, errorMessage }: GenericFallbackErrorProps) => {
  const { ssoEnabled } = useStores().config.getConfig();

  const goToDashboardLink = <ExternalLink href="/projects/dashboard"> </ExternalLink>;

  const internalError = !ssoEnabled && (errorStack || errorMessage);

  return (
    <Page qa="fallback-error" pageTitle={<PageTitle />}>
      <Section>
        <SimpleString qa="message">
          <Content value={x => x.pages.genericFallbackError.message} components={[goToDashboardLink]} />
        </SimpleString>
      </Section>

      {internalError && (
        <Section>
          <H2>
            <Content value={x => x.pages.genericFallbackError.developerHeading} />
          </H2>

          <pre style={{ whiteSpace: "pre-wrap" }}>{internalError}</pre>
        </Section>
      )}
    </Page>
  );
};
