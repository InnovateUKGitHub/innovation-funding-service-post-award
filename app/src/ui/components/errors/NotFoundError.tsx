import { PageTitle } from "@ui/features/page-title";

import { Page, Section } from "../layout";
import { SimpleString } from "../renderers/simpleString";
import { ExternalLink } from "../renderers";
import { Content } from "../content";

const innovateUKLink = "https://www.gov.uk/government/organisations/innovate-uk";

export const NotFoundError = () => {
  const dashboardLinkElement = (
    <a key="dashboard" href="/projects/dashboard">
      {" "}
    </a>
  );
  const innovateLinkElement = (
    <ExternalLink key="innovate" href={innovateUKLink}>
      {" "}
    </ExternalLink>
  );

  return (
    <Page qa="not-found" pageTitle={<PageTitle />}>
      <Section>
        <SimpleString qa="errorMessage">
          <Content
            value={x => x.pages.notFoundError.goBackMessage}
            components={[dashboardLinkElement, innovateLinkElement]}
          />
        </SimpleString>
      </Section>
    </Page>
  );
};
