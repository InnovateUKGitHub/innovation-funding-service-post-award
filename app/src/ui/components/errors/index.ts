export * from "./ErrorContainer";

// Unique Error UI's
export * from "./GenericFallbackError";
export * from "./UnauthenticatedError";
export * from "./NotFoundError";
