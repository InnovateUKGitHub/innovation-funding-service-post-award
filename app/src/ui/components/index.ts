import * as Renderers from "./renderers";
import * as Projects from "./projects";
import * as Claims from "./claims";
import * as Inputs from "./inputs";
import * as MonitoringReports from "./monitoringReports";

export * from "./accordion";
export * from "./accordionItem";
export * from "./details";
export * from "./documentList";
export * from "./documentSingle";
export * from "./errorSummary";
export * from "./form";
export * from "./layout";
export * from "./links";
export * from "./linksList";
export * from "./loading";
export * from "./logs";
export * from "./navigationArrows";
export * from "./partnersAndFinanceContacts";
export * from "./periodTitle";
export * from "./projectMember";
export * from "./styledButton";
export * from "./table";
export * from "./validationError";
export * from "./validationMessage";
export * from "./validationSummary";

export { Renderers, Projects, Claims, Inputs, MonitoringReports };
