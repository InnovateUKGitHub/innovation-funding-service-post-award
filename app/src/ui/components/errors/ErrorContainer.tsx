import { IAppError } from "@framework/types";
import { ErrorPayload, createErrorPayload } from "@shared/create-error-payload";
import { ContentProvider } from "@ui/redux/contentProvider";
import { PageTitleProvider } from "@ui/features/page-title";
import { useStores } from "@ui/redux";
import { MountedProvider } from "@ui/features";
import { useInitContent } from "@ui/features/use-initial-content";
import { FullHeight, GovWidthContainer, Header } from "@ui/components";
import { errorPages, internalErrorFallback, InternalErrorTypes } from "./error.config";

export type ErrorContainerProps = ErrorPayload["params"] & { from?: string };

/**
 * Error hoc which switches between UI based on errorType
 */
export const ErrorContainer = (props: ErrorContainerProps) => {
  const predefinedError = errorPages[props.errorType as InternalErrorTypes];

  const ErrorUI = predefinedError || internalErrorFallback;

  return <ErrorUI {...props} />;
};

/**
 * Pull required props from ErrorBoundary
 * Requires Providers and page layout because fallback lies outside main react component tree
 */
export function ErrorBoundaryFallback({ error }: { error: IAppError }) {
  const stores = useStores();
  const errorPayload = createErrorPayload(error, false);
  const content = useInitContent();
  const config = stores.config.getConfig();
  return (
    <MountedProvider>
      <ContentProvider value={content}>
        <PageTitleProvider title="Sorry, there is a problem with the service">
          <FullHeight.Container>
            <a href="#main-content" className="govuk-skip-link">
              Skip to main content
            </a>
            <Header headingLink={`${config.ifsRoot}/competition/search`} />
            <FullHeight.Content>
              <GovWidthContainer>
                <ErrorContainer {...errorPayload.params} />
              </GovWidthContainer>
            </FullHeight.Content>
          </FullHeight.Container>
        </PageTitleProvider>
      </ContentProvider>
    </MountedProvider>
  );
}
