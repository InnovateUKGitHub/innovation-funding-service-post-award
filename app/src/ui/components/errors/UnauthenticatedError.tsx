import { PageTitle } from "@ui/features/page-title";
import { Page, Section, Content } from "@ui/components";
import { ExternalLink, SimpleString } from "@ui/components/renderers";
import { UserChanger } from "@ui/containers/developer/UserChanger";
import { useStores } from "@ui/redux";

export function UnauthenticatedError() {
  const { ssoEnabled } = useStores().config.getConfig();

  const contactLink = "https://apply-for-innovation-funding.service.gov.uk/info/contact";

  return (
    <Page qa="unauthenticated-error" pageTitle={<PageTitle />}>
      <Section>
        <SimpleString qa="reason">
          <Content
            value={x => x.pages.unauthenticatedError.contactUs}
            components={[
              <ExternalLink key="link" href={contactLink}>
                {" "}
              </ExternalLink>,
            ]}
          />
        </SimpleString>
      </Section>

      {!ssoEnabled && <UserChanger />}
    </Page>
  );
}
