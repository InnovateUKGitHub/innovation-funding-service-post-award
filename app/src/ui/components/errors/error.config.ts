import React from "react";
import { UnauthenticatedError } from "./UnauthenticatedError";
import { GenericFallbackError } from "./GenericFallbackError";
import { NotFoundError } from "./NotFoundError";

/**
 * @description Treat error messaging as unique key lookup, the type should be string.
 *
 * To ensure that we can scale the error handling, there maybe keys we do not have content for... In this we have a fallback
 */
export type InternalErrorTypes = "AUTHENTICATION_ERROR" | "NOT_FOUND";
export type ErrorTypes = InternalErrorTypes & "INTERNAL_ERROR_FALLBACK";

type ErrorConfig = Record<InternalErrorTypes, React.ElementType>;

export const internalErrorFallback: typeof GenericFallbackError = GenericFallbackError;

export const errorPages: ErrorConfig = {
  AUTHENTICATION_ERROR: UnauthenticatedError,
  NOT_FOUND: NotFoundError,
};
