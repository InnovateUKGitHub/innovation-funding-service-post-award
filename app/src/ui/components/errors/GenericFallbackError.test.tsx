import { render } from "@testing-library/react";

import { TestBed } from "@shared/TestBed";
import { GenericFallbackError, GenericFallbackErrorProps } from "@ui/components/errors";
import { ErrorCode } from "@framework/types";
import { testInitialiseInternationalisation } from "@shared/testInitialiseInternationalisation";

describe("<GenericFallbackError />", () => {
  describe("@renders", () => {
    const stubContent = {
      pages: {
        genericFallbackError: {
          message: "Please use <0>Google</0>.",
        },
      },
    };

    const setup = (props: GenericFallbackErrorProps) =>
      render(
        <TestBed>
          <GenericFallbackError {...props} />
        </TestBed>,
      );

    beforeAll(async () => {
      await testInitialiseInternationalisation(stubContent);
    });

    test("with page message containing link text", () => {
      const { queryByTestId } = setup({
        errorCode: ErrorCode.UNKNOWN_ERROR,
        errorType: "stub-errorType",
      });

      const message = queryByTestId("message");

      expect(message).toBeInTheDocument();
      expect(message).toMatchSnapshot();
    });

    describe("with developer error", () => {
      test("without errorMessage", () => {
        const { container } = setup({
          errorCode: ErrorCode.UNKNOWN_ERROR,
          errorType: "stub-errorType",
        });

        expect(container.querySelector("pre")).not.toBeInTheDocument();
      });

      test("with errorMessage", () => {
        const stubDeveloperError = "stub-developer-error";

        const { queryByText } = setup({
          errorCode: ErrorCode.UNKNOWN_ERROR,
          errorType: "stub-errorType",
          errorMessage: stubDeveloperError,
        });

        expect(queryByText(stubDeveloperError, { selector: "pre" })).toBeInTheDocument();
      });
    });
  });
});
