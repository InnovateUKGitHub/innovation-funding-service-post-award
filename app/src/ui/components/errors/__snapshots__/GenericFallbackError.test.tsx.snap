// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenericFallbackError /> @renders with page message containing link text 1`] = `
<p
  class="govuk-body"
  data-qa="message"
>
  Please use 
  <a
    href="/projects/dashboard"
    rel="noopener noreferrer"
    target="_blank"
  >
    Google
  </a>
  .
</p>
`;
